### PR TITLE
Add support of macrobenchmark result analysis in fireci

### PR DESCRIPTION
**Pull request by yifanyang**

### CODE REVIEW

**Review by rlazo — APPROVED** [review]

File: ci/fireci/fireciplugins/macrobenchmark/analyze/aggregator.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  # Calculate dispersions of each percentile over all runs
+  mean = percentiles.mean()
+  std = percentiles.std()  # standard deviation
+  cv = std / mean  # coefficient of variation (relative standard deviation)
+  mad = (percentiles - percentiles.mean()).abs().mean()  # mean absolute deviation
+  rmad = mad / mean  # relative mean absolute deviation (mad / mean)
+  dispersions = pd.DataFrame([pd.Series(cv, name='cv'), pd.Series(rmad, name='rmad')])
+
+  # Optionally save percentiles and dispersions to file
+  if output_dir is not None:
```
[the remark this answers is not in the document]
Comment (yifanyang) [reply ci/fireci/fireciplugins/macrobenchmark/analyze/aggregator.py]: Done

File: ci/fireci/fireciplugins/macrobenchmark/analyze/analyzer.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    local_reports_dir: Path,
+    ctl_ftl_results_dir: list[str],
+    ctl_local_reports_dir: Path,
+    exp_ftl_results_dir: list[str],
+    exp_local_reports_dir: Path,
+    output_dir: Path
+):
+  logger.info('Starting to analyze macrobenchmark test results ...')
+
+  if output_dir is None:
```
[the remark this answers is not in the document]
Comment (yifanyang) [reply ci/fireci/fireciplugins/macrobenchmark/analyze/analyzer.py]: Done

File: ci/fireci/fireciplugins/macrobenchmark/analyze/utils.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+logger = logging.getLogger('fireci.macrobenchmark')
+DataPoint = TypedDict('DataPoint', {'duration': float, 'device': str, 'trace': str, 'run_id': str})
+
+
+def collect_data_points(ftl_results_dir: list[str], local_reports_dir: Path) -> list[DataPoint]:
+  if not ftl_results_dir and local_reports_dir is None:
+    raise ClickException('Neither ftl-results-dir or local-reports-dir is provided.')
+  elif ftl_results_dir and local_reports_dir is None:
```
[the remark this answers is not in the document]
Comment (yifanyang) [reply ci/fireci/fireciplugins/macrobenchmark/analyze/utils.py]: Done

File: ci/fireci/fireciplugins/macrobenchmark/analyze/aggregator.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  exp_percentiles, _ = calculate_statistic(trace, device, experimental)
+
+  ctl_mean = ctl_percentiles.mean()
+  exp_mean = exp_percentiles.mean()
+
+  delta = exp_mean - ctl_mean
+  percentage = delta / ctl_mean
+
+  # Optionally save statistics to file
+  if output_dir is not None:
```
[the remark this answers is not in the document]
Comment (yifanyang) [reply ci/fireci/fireciplugins/macrobenchmark/analyze/aggregator.py]: Done

File: ci/fireci/fireciplugins/macrobenchmark/analyze/utils.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+from pathlib import Path
+from typing import TypedDict
+
+
+logger = logging.getLogger('fireci.macrobenchmark')
+DataPoint = TypedDict('DataPoint', {'duration': float, 'device': str, 'trace': str, 'run_id': str})
+
+
+def collect_data_points(ftl_results_dir: list[str], local_reports_dir: Path) -> list[DataPoint]:
+  if not ftl_results_dir and local_reports_dir is None:
```
[the remark this answers is not in the document]
Comment (yifanyang) [reply ci/fireci/fireciplugins/macrobenchmark/analyze/utils.py]: Done

File: ci/fireci/fireciplugins/macrobenchmark/analyze/utils.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    raise ClickException('Neither ftl-results-dir or local-reports-dir is provided.')
+  elif ftl_results_dir and local_reports_dir is None:
+    temp_dir = _download(ftl_results_dir)
+    return _extract_raw_data(temp_dir)
+  elif not ftl_results_dir and local_reports_dir is not None:
```
[the remark this answers is not in the document]
Comment (yifanyang) [reply ci/fireci/fireciplugins/macrobenchmark/analyze/utils.py]: Done

**Pull request merged**